[CHIA-710] Add the concept of 'action scopes'

Quexington opened the pull request:
This PR creates a new utility called `ActionScope`.  The motivation for this originated in the wallet, but this utility could be more broadly applicable.  The purpose is to provide a standard way for multiple functions/coroutines to access and modify some shared state as part of a single transaction.

When an action scope is created the scope is passed from function to coroutine etc.  and if a logical process wishes to update or read from the state, it must "check out" the state and return it when it is done. No other process can check out the state while it is checked out by another.  Processes may also register callbacks to be run when the action scope is about to close.  Right before the action scope is closed, these callbacks run and then the state is deleted.  Other layers on top can choose to commit that state or whatever else if they so wish.

The wallet motivation more specifically is to provide a single hub for all of the state changes that will happen as part of creating a transaction.  Right now, every function is responsible for just making side effects at will and some higher process is responsible for making sure these side effects don't happen prematurely and get rolled back if something goes wrong.  Every step along the way needs to make sure the information is sent up and down properly so that everything works properly.  It's quite a burden.  With action scopes, everything just modifies this temporary state and then the thing that made the action scope can decide what to do with everything when it closes.  In addition, every function that is called can see the _whole_ context it is operating in rather than just what its caller has told it.  This will be important for vault singletons later on.

This may sound very similar to the `DBWrapper` functionality.  You would be correct in thinking that as that is what action scopes use as a back end.  It's not clear we want to always use an in memory sqlite DB but it's the fastest path to getting all of these features.  In any case, the backend is a protocol that can be switched out at a later date, even if only in some areas.

Quexington committed:
Add the concept of 'action scopes'

Quexington committed:
pylint and test coverage

Quexington committed:
add try/finally

altendky commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+        async with DBWrapper2.managed(":memory:", reader_count=0) as db:
+            self = cls(db)
+            async with self._db.writer() as conn:
+                await conn.execute("CREATE TABLE memos(" " key blob," " value blob" ")")
+                await conn.execute("CREATE TABLE side_effects(" " total blob" ")")
```
```suggestion
                await conn.execute("CREATE TABLE memos(key blob, value blob)")
                await conn.execute("CREATE TABLE side_effects(total blob)")
```

Quexington replied:
Taken: ef59007507

altendky commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+        return side_effects
+
+    async def save_resource(self, resource: SideEffects) -> None:
+        await self.get_active_writer().execute("DELETE FROM side_effects")
+        await self.get_active_writer().execute(
+            "INSERT INTO side_effects VALUES(?)",
+            (bytes(resource),),
+        )
```
Maybe worth using the upsert alternative to make this a single query?  Though sure, theoretical optimization and all...  If interested, you or I will have to lookup the form.

Quexington replied:
Not sure what you mean so... no? :)

altendky replied:
https://www.sqlite.org/lang_upsert.html

Quexington replied:
Changed to just always use UPDATE: 94b2d65f53

altendky commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+    ) -> AsyncIterator[ActionScope[_T_SideEffects]]:
+        async with resource_manager_backend.managed(side_effects_format()) as resource_manager:
+            self = cls(_resource_manager=resource_manager, _side_effects_format=side_effects_format)
+            try:
+                yield self
+            except Exception:
+                raise
+            else:
+                async with self.use(_callbacks_allowed=False) as interface:
+                    for callback in self._callbacks:
+                        await callback(interface)
+                    self._final_side_effects = interface.side_effects
```
Is there really nothing to clean up in the case of an exception?

Just from a language mechanics perspective, this can be simplified.

```suggestion
            yield self

            async with self.use(_callbacks_allowed=False) as interface:
                for callback in self._callbacks:
                    await callback(interface)
                self._final_side_effects = interface.side_effects
```

Quexington replied:
No, this is all temporary state while it's active. Only once everything finishes do you decide to commit it if you want.  Your suggestion has been taken: ef59007507

Quexington replied:
Oh actually pylint didn't like this one either

Quexington replied:
Changed back pending pylint fix in main: 94b2d65f53

altendky commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+                raise
+            else:
+                async with self.use(_callbacks_allowed=False) as interface:
+                    for callback in self._callbacks:
+                        await callback(interface)
```
As in the overall review comment, what considerations have there been around callback ordering and response to failures?

Quexington replied:
Callbacks no longer have ordering

altendky commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+            memos = await self._resource_manager.get_memos()
+            side_effects = await self._resource_manager.get_resource(self._side_effects_format)
+            interface = StateInterface(memos, side_effects, _callbacks_allowed)
+            try:
+                yield interface
+            except Exception:
+                raise
+            else:
+                await self._resource_manager.save_memos(interface.memos)
+                await self._resource_manager.save_resource(interface.side_effects)
+                self._callbacks.extend(interface._new_callbacks)
```
```suggestion
            yield interface

            await self._resource_manager.save_memos(interface.memos)
            await self._resource_manager.save_resource(interface.side_effects)
            self._callbacks.extend(interface._new_callbacks)
```

Quexington replied:
Interestingly pylint complains about this one.  I don't fully understand it tbh: `chia\util\action_scope.py:142:12: W0135: The context used in function 'new_scope' will not be exited. (contextmanager-generator-missing-cleanup)`

Quexington replied:
Changed back pending pylint fix in main: 94b2d65f53

altendky reviewed (changes requested):
What affect does an error in a callback have on a transaction?  I think it's complex because what if the first callback goes through (and this is where side effects would expect to get implemented?) and the second fails?  How do we recover?

Relatedly, there's the option of concurrent callbacks (good if they are independent, bad if they depend on order which maybe is questionable anyways if they are added by code in unrelated places which is unaware of the other code) or just callback order in general, and also the idea of exception groups since we have multiple things to do.

Whatever should happen, there should be a test for nested `.use()` managers.  I believe that is allowed presently due to the way the two-layers-down `DBWrapper2` works.

Quexington committed:
Address comments by @altendky

Quexington committed:
Address comments by @altendky

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+    state_interface = StateInterface({}, TestSideEffects(), True)
+    initial_count = len(state_interface._new_callbacks)
+    state_interface.add_callback(default_async_callback)
+    assert len(state_interface._new_callbacks) == initial_count + 1
```
Taken and extended a bit: a50dd8f96a

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+        yield scope
+
+
+# Test creating a new ActionScope and ensuring tables are created
```
Not sure, I'll change it to something more sensical: a50dd8f96a

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+
+    async with action_scope.use() as interface:
+        assert interface.memos[b"foo"] == b"bar"
+        assert interface.side_effects.buf == b"bar"
```
Good suggestion, taken: a50dd8f96a

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+            interface.side_effects.buf = b"qat"
+            raise RuntimeError("Going to be caught")
+    except RuntimeError:
+        pass
```
Oh yeah I'm dumb (fixed the one below too): a50dd8f96a

Quexington commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+_T_SideEffects = TypeVar("_T_SideEffects", bound=SideEffects)
+
+
+@dataclass
```
Taken: ef59007507

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+
+                interface.add_callback(callback)
+
+            async with action_scope.use() as interface:
+                raise RuntimeError("This should prevent the callbacks from being called")
```
Sure, makes enough sense to me: a50dd8f96a

Quexington commented on chia/util/action_scope.py:
```diff
@@ -0,0 +1,174 @@
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def managed(cls, initial_resource: SideEffects) -> AsyncIterator[ResourceManager]:  # pragma: no cover
+        # We have to put this yield here for mypy to recognize the function as a generator
```
Taken under protest (protocols are a typing construct and the typing should be smart enough to understand that the `yield` is in the `...` just like it takes `return` for granted in the same way) (resolved dup below): ef59007507

Quexington commented on chia/_tests/util/test_action_scope.py:
```diff
@@ -0,0 +1,118 @@
+
+# Fixture to create an ActionScope with a mocked DBWrapper2
+@pytest.fixture
+async def action_scope() -> AsyncIterator[ActionScope[TestSideEffects]]:
```
Smart & taken: a50dd8f96a

Quexington committed:
86 memos

Quexington committed:
Only one callback

Quexington commented:
I killed memos because implementing this upstream I didn't use them (I will eventually) and they can just be part of side effects.  Separating them out was pointless.  I also removed the ability to have a list of callbacks due to the concerns around ordering.  There's no good solution for that, so instead, I changed the paradigm so that people _have_ to know about what other callbacks they are potentially interfering with (because there's exactly one place where they can all live).

Quexington committed:
pylint

Quexington committed:
Address more comments by @altendky

Quexington committed:
remove unused variable

altendky reviewed (changes requested):
chatted with quex, we need a better understanding around what is being updated, do we accidentally have multiple rows?  etc.

Quexington committed:
add comment

markelrod commented:
coverage exceptions are fine
